assets/src/components/detours/diversionPage.tsx

```diff
@@ -1,7 +1,107 @@
-import React from "react"
+import React, { useEffect, useMemo, useState } from "react"
 import { DiversionPanel, DiversionPanelProps } from "./diversionPanel"
 import { DetourMap } from "./detourMap"
-import { Shape } from "../../schedule"
+import { Shape, ShapePoint } from "../../schedule"
+import { LatLngLiteral } from "leaflet"
+import { fetchDetourDirections } from "../../api"
+import {
+  latLngLiteralToShapePoint,
+  shapePointToLatLngLiteral,
+} from "../../util/pointLiterals"
+
+const useDetourDirections = (shapePoints: ShapePoint[]) => {
+  const [detourShape, setDetourShape] = useState<LatLngLiteral[]>([])
+
+  useEffect(() => {
+    let shouldUpdate = true
+
+    fetchDetourDirections(shapePoints).then((detourShape) => {
+      if (detourShape && shouldUpdate) {
+        setDetourShape(detourShape.coordinates.map(shapePointToLatLngLiteral))
+      }
+    })
+
+    return () => {
+      shouldUpdate = false
+    }
+  }, [shapePoints])
+
+  return detourShape
+}
+
+const useDetour = () => {
+  const [startPoint, setStartPoint] = useState<LatLngLiteral | null>(null)
+  const [endPoint, setEndPoint] = useState<LatLngLiteral | null>(null)
+  const [waypoints, setWaypoints] = useState<LatLngLiteral[]>([])
+
+  const detourShape = useDetourDirections(
+    [startPoint, ...waypoints, endPoint]
+      .filter((v): v is LatLngLiteral => !!v)
+      .map(latLngLiteralToShapePoint)
+  )
```

Comment:
```suggestion
  const allPoints = useMemo(
    () =>
      [startPoint, ...waypoints, endPoint]
        .filter((v): v is LatLngLiteral => !!v)
        .map(latLngLiteralToShapePoint),
    [startPoint, endPoint, waypoints]
  )

  const detourShape = useDetourDirections(allPoints)
```

I think this fixes the infinite API call issue above. (You've unleashed a monster! A memoized monster! So I guess only one monster at a time unless parameters change.)

Reply:
I have no attachment at all to the name `allPoints`, fwiw

Reply:
Thanks! 84afffa8